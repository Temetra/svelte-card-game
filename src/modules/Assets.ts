export * from "@/modules/Assets/Audio";
export * from "@/modules/Assets/CardImages";
export * from "@/modules/Assets/LoadingStatus";
export * from "@/modules/Assets/Text";