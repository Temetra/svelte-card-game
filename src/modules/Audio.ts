import { checkResponse } from "./Fetching";

const AudioContext = window.AudioContext || window.webkitAudioContext;
const audioCtx = new AudioContext();
const sounds: Partial<Record<string, AudioBuffer>> = {};

export function prepareAudio() {
	return Promise.all([
		loadAudio("sounds/blip.mp3"),
		loadAudio("sounds/card-hard.mp3"),
		loadAudio("sounds/card-soft.mp3")
	])
	.then(values => {
		sounds.blip = values[0];
		sounds.card = values[1];
		sounds.slide = values[2];
	});
}

function loadAudio(filename: string): Promise<AudioBuffer> {
	return fetch(filename)
		.then(checkResponse)
		.then(response => response.arrayBuffer())
		.then(arrayBuffer => audioCtx.decodeAudioData(arrayBuffer));
}

export function playSound(name: string) {
	let buffer = sounds[name];

	if (buffer) {
		const trackSource = audioCtx.createBufferSource();
		trackSource.buffer = buffer;
		trackSource.connect(audioCtx.destination);
		trackSource.start();
	}
}
