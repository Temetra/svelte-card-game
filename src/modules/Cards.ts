export * from "@/modules/Cards/Card";
export * from "@/modules/Cards/CardState";
export * from "@/modules/Cards/GetCardName";
export * from "@/modules/Cards/Rank";
export * from "@/modules/Cards/Suit";
export * from "@/modules/Cards/Scoring";
