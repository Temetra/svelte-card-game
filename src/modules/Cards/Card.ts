import type { Rank } from "@/modules/Cards/Rank";
import type { Suit } from "@/modules/Cards/Suit";
import type { CardState } from "@/modules/Cards/CardState";

export interface Card {
	rank: Rank,
	suit: Suit
}

export interface StatefulCard extends Card {
	state: CardState
}

export function getCardIndex(card: Card);
export function getCardIndex(rank: Rank, suit: Suit);
export function getCardIndex(x: any, y?: any) {
	if (typeof x == "number" && y != null) {
		// Assume func was given rank and suit
		return y * 100 + x;
	}
	else if (typeof x == "object") {
		// Assume x is a card object
		return x.suit * 100 + x.rank;
	}

	return null;
}