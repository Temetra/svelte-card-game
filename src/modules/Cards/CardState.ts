export enum CardState {
	Hidden = 1 << 0,
	Focused = 1 << 1,
	Flipped = 1 << 2,
	Dealing = 1 << 3,
	Spinning = 1 << 4,
	Discarding = 1 << 5,
}