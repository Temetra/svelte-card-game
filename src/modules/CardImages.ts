import { Suit, Rank, Suits, Ranks } from "./Cards";
import { fetchFiles } from "./Fetching";
import { loading } from "@/stores/datastore";

// In-memory cache for card images
let images: Record<string, string>;

// Returns promise to preload images into bank
export function prepareCardGraphics() {
	// Create an array of images to load
	const files = {
		"CardBackground": "images/cards/CardBackground.svg",
		"Joker": "images/cards/Joker.svg"
	};
	
	// Iterate over card enums, adding a card for each combo
	for (const suit of Suits) {
		for (const rank of Ranks) {
			let name = `${Suit[suit]}${Rank[rank]}`;
			files[name] = `images/cards/${name}.svg`;
		}
	}

	// Return a promise to load images
	return fetchFiles(files, processResponse, updateProgress)
		.then(data => images = data);
}

// Get blob from response, convert to base64
async function processResponse(response: Response): Promise<string> {
	return await response.blob()
		.then(blob => new Promise<string>(resolve => {
			const reader = new FileReader();
			reader.addEventListener("error", () => resolve(null));
			reader.addEventListener("load", evt => resolve(evt.target.result as string));
			reader.readAsDataURL(blob);
		}));
}

// Update progress store
function updateProgress(progress: number, total: number) {
	loading.set({ name:"images", progress, total });
}

// Returns base64 image blob for card face background
export function getCardFace() : string {
	return images["CardBackground"];
}

// Returns base64 image blob for card face value
export function getCardValue(suit: Suit, rank: Rank) : string {
	switch (suit) {
		case Suit.Joker: return images["Joker"];
		default: return images[`${Suit[suit]}${Rank[rank]}`];
	}
}
