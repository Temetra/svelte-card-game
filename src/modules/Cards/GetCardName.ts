import { Rank } from "@/modules/Cards/Rank";
import { Suit } from "@/modules/Cards/Suit";

export function getCardName(suit: Suit, rank: Rank) {
	if (suit == Suit.Joker) return Suit[suit];
	else return `${Rank[rank]} of ${Suit[suit]}`;
}
