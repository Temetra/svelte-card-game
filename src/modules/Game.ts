import { Card, State } from "./Cards";
import { Suits, Ranks, State as CardState } from "./Cards";
import { waitFor } from "./Fetching";
import { playSound } from "./Audio";
import { controlsEnabled, deck, playerOneHand } from "@/stores/datastore";

export async function prepareGame() {
	controlsEnabled.set(true);
}

export function requestNewCards() {
	return startDealing()
		.then(dealAllCards)
		.then(animateDealtCard)
		.then(animateDealtCard)
		.then(animateDealtCard)
		.then(animateDealtCard)
		.then(animateDealtCard)
		.then(finishDealing);
}

async function startDealing() {
	// Disable button
	controlsEnabled.set(false);

	// Audio feedback
	playSound("blip");

	// Clear player hand
	playerOneHand.set([]);

	// Some sort of anim prep before dealing
	await waitFor(500);
	resetDeck();
	shuffleDeck();
}

async function dealAllCards() {
	// Array to store drawn cards
	let drawn: Card[];

	// Draw 5 cards
	deck.update(cards => {
		drawn = cards.splice(0, 5);
		return cards;
	});

	// Change state of cards
	drawn.map(card => card.state = CardState.Hidden);

	// Add card to hand
	playerOneHand.update(hand => {
		hand.push(...drawn);
		return hand;
	});
}

async function animateDealtCard() {
	const avgDelay = 400;
	const delayDeviation = 25;
	const audioDelay = 0;
	const delayFrom = avgDelay - delayDeviation - audioDelay;
	const delayTo = avgDelay + delayDeviation - audioDelay;

	// Delay before dealing
	await waitFor(randomFromRange(delayFrom, delayTo));

	// Find first hidden card in hand and change state
	playerOneHand.update(hand => {
		hand.find(x => x.state === State.Hidden).state = State.Dealing;
		return hand;
	});

	// Audio feedback
	await waitFor(audioDelay);
	playSound("card")
}

async function finishDealing() {
	// Enable button
	await waitFor(500);
	controlsEnabled.set(true);
}

function randomFromRange(min: number, max: number) : number {
	return Math.random() * (max - min) + min;
}

function resetDeck() {
	// Create new array
	const result: Card[] = [];
	const state = CardState.Default;

	// Iterate over suits and ranks, adding to array
	for (const suit of Suits) {
		for (const rank of Ranks) {
			result.push({ suit, rank, state });
		}
	}

	// Set deck store to array
	deck.set(result);
}

function shuffleDeck() {
	deck.update(array => {
		let currentIndex = array.length;
		let temporaryValue: Card;
		let randomIndex: number;
		
		while (currentIndex != 0) {
			randomIndex = Math.floor(Math.random() * currentIndex);
			currentIndex -= 1;
			temporaryValue = array[currentIndex];
			array[currentIndex] = array[randomIndex];
			array[randomIndex] = temporaryValue;
		}
		
		return array;
	});
}
