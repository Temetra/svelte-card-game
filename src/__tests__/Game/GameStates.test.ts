import { GameState, gameState, playerOneHand, requestNewCards, discardAndDraw } from "@/modules/Game";
import { CardState } from "@/modules/Cards";

let states: string[] = [];
let unsubGameState;

let hands: string[][] = [];
let unsubHands;

beforeEach(() => {
	if (unsubGameState) unsubGameState();
	gameState.set(GameState.Preparing);
	states = [];
	unsubGameState = gameState.subscribe(x => states.push(GameState[x]));

	if (unsubHands) unsubHands();
	playerOneHand.set([]);
	hands = [];
	unsubHands = playerOneHand.subscribe(x => hands.push(x.map(y => CardState[y.state])));
});

test("requestNewCards", async () => {
	expect(states).toStrictEqual(["Preparing"]);
	expect(hands).toStrictEqual([[]]);
	
	await requestNewCards();
	
	expect(states).toMatchSnapshot();
	expect(hands).toMatchSnapshot();
});

test("DiscardAndDraw", async () => {
	expect(states).toStrictEqual(["Preparing"]);
	expect(hands).toStrictEqual([[]]);
	
	await requestNewCards();

	playerOneHand.update(hand => {
		hand[0].state = CardState.Flipped;
		hand[3].state = CardState.Flipped;
		return hand;
	});

	await discardAndDraw();
	
	expect(states).toMatchSnapshot();
	expect(hands).toMatchSnapshot();
});