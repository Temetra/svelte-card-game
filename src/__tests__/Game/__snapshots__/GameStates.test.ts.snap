// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiscardAndDraw 1`] = `
Array [
  "Preparing",
  "Dealing",
  "Selection",
  "Discard",
  "Drawing",
  "Summary",
]
`;

exports[`DiscardAndDraw 2`] = `
Array [
  Array [],
  Array [],
  Array [
    "Hidden",
    "Hidden",
    "Hidden",
    "Hidden",
    "Hidden",
  ],
  Array [
    "Dealing",
    "Hidden",
    "Hidden",
    "Hidden",
    "Hidden",
  ],
  Array [
    "Dealing",
    "Dealing",
    "Hidden",
    "Hidden",
    "Hidden",
  ],
  Array [
    "Dealing",
    "Dealing",
    "Dealing",
    "Hidden",
    "Hidden",
  ],
  Array [
    "Dealing",
    "Dealing",
    "Dealing",
    "Dealing",
    "Hidden",
  ],
  Array [
    "Dealing",
    "Dealing",
    "Dealing",
    "Dealing",
    "Dealing",
  ],
  Array [
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  Array [
    "Flipped",
    undefined,
    undefined,
    "Flipped",
    undefined,
  ],
  Array [
    "Discarding",
    undefined,
    undefined,
    "Discarding",
    undefined,
  ],
  Array [
    "Dealing",
    undefined,
    undefined,
    "Discarding",
    undefined,
  ],
  Array [
    "Dealing",
    undefined,
    undefined,
    "Dealing",
    undefined,
  ],
  Array [
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
  ],
]
`;

exports[`requestNewCards 1`] = `
Array [
  "Preparing",
  "Dealing",
  "Selection",
]
`;

exports[`requestNewCards 2`] = `
Array [
  Array [],
  Array [],
  Array [
    "Hidden",
    "Hidden",
    "Hidden",
    "Hidden",
    "Hidden",
  ],
  Array [
    "Dealing",
    "Hidden",
    "Hidden",
    "Hidden",
    "Hidden",
  ],
  Array [
    "Dealing",
    "Dealing",
    "Hidden",
    "Hidden",
    "Hidden",
  ],
  Array [
    "Dealing",
    "Dealing",
    "Dealing",
    "Hidden",
    "Hidden",
  ],
  Array [
    "Dealing",
    "Dealing",
    "Dealing",
    "Dealing",
    "Hidden",
  ],
  Array [
    "Dealing",
    "Dealing",
    "Dealing",
    "Dealing",
    "Dealing",
  ],
  Array [
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
  ],
]
`;
